// @roots/entrypoints-webpack-plugin/lib/cjs/webpack.plugin.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntrypointsWebpackPlugin = void 0;
const autobind_decorator_1 = require("autobind-decorator");
const lodash_1 = require("lodash");
const webpack_1 = require("webpack");
const inline_emitter_1 = require("./inline.emitter");
/**
 * Produces `entrypoints.json` artifact with compiled assets broken down
 * by entrypoint and then filetype.
 *
 * @example
 * ```
 * import {EntrypointsWebpackPlugin} from '@roots/entrypoints-webpack-plugin'
 *
 * const config = {
 *   plugins: [new EntrypointsWebpackPlugin()]
 * }
 * ```
 *
 * @public
 */
class EntrypointsWebpackPlugin {
    /**
     * Class constructor
     *
     * @public
     */
    constructor(options) {
        this.options = options;
        /**
         * Plugin compiler ident
         *
         * @public
         */
        this.plugin = {
            name: 'EntrypointsManifestPlugin',
            stage: Infinity,
        };
        /**
         * Artifact filename
         *
         * @public
         */
        this.name = 'entrypoints.json';
    }
    /**
     * Webpack plugin API's `apply` hook
     *
     * @public
     * @decorator `@bind`
     */
    apply(compiler) {
        this.assets = {};
        this.compiler = compiler;
        this.options.publicPath =
            this.options.publicPath ??
                this.compiler.options.output.publicPath ??
                '';
        this.options.publicPath = this.options.publicPath.replace('auto', '');
        this.compiler.hooks.thisCompilation.tap(this.plugin, (compilation) => {
            this.compilation = compilation;
            this.compilation.hooks.processAssets.tap({ ...this.plugin, additionalAssets: true }, this.processAssets);
        });
    }
    /**
     * Runs through each entrypoint entry and adds to the
     * manifest
     *
     * @public
     * @decorator `@bind`
     */
    processAssets() {
        this.compilation.entrypoints.forEach(entry => {
            this.getEntrypointFiles(entry)
                .filter(({ file }) => !file.includes('hot-update'))
                .map(({ key, file }) => {
                this.addToManifest({ key, entry: entry.name, file });
            });
        });
        this.options.emitHtml &&
            new inline_emitter_1.InlineEmitter(this.compilation, this.assets, this.options.publicPath).emitHtmlTags();
        Object.assign(this.compilation.assets, {
            [this.name]: new webpack_1.sources.RawSource(JSON.stringify(this.assets), true),
        });
    }
    /**
     * Adds an entry to the manifest
     *
     * @public
     * @decorator `@bind`
     */
    addToManifest({ key = null, entry, file }) {
        const type = file.split('.').pop();
        if (!this.assets[entry]) {
            this.assets[entry] = {
                [type]: null,
            };
        }
        this.assets[entry] = {
            ...this.assets[entry],
            [type]: this.options.type === 'object' && key
                ? {
                    ...(this.assets[entry][type] ?? {}),
                    [key]: this.options.publicPath.concat(file),
                }
                : (0, lodash_1.uniq)([
                    ...(this.assets[entry][type] ?? []),
                    this.options.publicPath.concat(file),
                ]),
        };
    }
    /**
     * Get assets from an entrypoint
     *
     * @public
     * @decorator `@bind`
     */
    getEntrypointFiles(entry) {
        const files = [];
        for (const chunk of entry.chunks) {
            Array.from(chunk.files).map(file => {
                files.push({
                    key: chunk.name,
                    file,
                });
            });
        }
        return files;
    }
}
__decorate([
    autobind_decorator_1.boundMethod
], EntrypointsWebpackPlugin.prototype, "apply", null);
__decorate([
    autobind_decorator_1.boundMethod
], EntrypointsWebpackPlugin.prototype, "processAssets", null);
__decorate([
    autobind_decorator_1.boundMethod
], EntrypointsWebpackPlugin.prototype, "addToManifest", null);
__decorate([
    autobind_decorator_1.boundMethod
], EntrypointsWebpackPlugin.prototype, "getEntrypointFiles", null);
exports.EntrypointsWebpackPlugin = EntrypointsWebpackPlugin;
